symbolic-base/test/Tests/ReedSolomon.hs

```diff
@@ -0,0 +1,111 @@
+  arbitrary = do
+    let pe = fromConstant (3 :: Natural)
+        c = fromIntegral $ min 100 (order @f)
+    t <- chooseInt (1,P.div c 3)
+    let r = 2*t
+        k = c P.- r P.- 1
+    msg <- vector @f k
+    er <- polyErr (k+t) t
+    return $ ReedSolomonExample pe k r msg er
+
+polyErr :: forall c. (Arbitrary c, Field c) => Int -> Int -> Gen [c]
+polyErr kt t = do
+    pos <- shuffle (replicate t (1 :: Integer) P.++ replicate kt 0)
+    mapM (\p -> bool arbitrary (pure zero) (p==0) ) pos
+
+----------------------------------------------------------------------------------------
+
+propGenerator :: forall c. (FiniteField c, Eq c) => ReedSolomonExample c -> Bool
+propGenerator ReedSolomonExample {..} =
+    let vals = take r $ iterate (C.* pe) (pe :: c)
+        polyGen = generator r pe
+    in all (\x -> evalPoly polyGen x == zero) vals
+
+propEncoder :: forall c. (Field c, Eq c) => ReedSolomonExample c -> Bool
+propEncoder ReedSolomonExample {..} =
+    let encodedMsg = encode msg pe r
+        reminder = snd $ qr encodedMsg (generator r pe)
+    in deg reminder == -1
+
+propBerlekampNoError :: forall c. (FiniteField c, Eq c) => ReedSolomonExample c -> Bool
+propBerlekampNoError ReedSolomonExample {..} =
+    let
+        vals = V.iterateN r (C.* pe) pe
+        encodedMsg = encode msg pe r
+        syndromes = toPoly $ V.map (evalPoly encodedMsg) vals
+    in syndromes == zero && berlekamp syndromes r == (0, one)
+
+propBerlekampWithErrors :: forall c. (FiniteField c, Ord c) => ReedSolomonExample c -> Bool
+propBerlekampWithErrors ReedSolomonExample {..} =
+    let
+        vals = V.iterateN r (C.* pe) pe
+        encoded' = encode msg pe r
+        errorMsg = toPoly $ V.fromList err
+        encoded = encoded' C.+ errorMsg
+        syndromes = toPoly $ V.map (evalPoly encoded) vals
+        bl = snd $ berlekamp syndromes r
+        roots = filter (/= zero) $ map (\x -> bool zero x (evalPoly bl x == zero)) $ takeWhile (/= pe) $ iterate (C.* pe) pe
+
+        es1 = takeWhile (/= one) $ iterate (C.* pe) one
+        rightLocators = filter (/= zero) $ zipWith (\e p -> bool zero (finv p) (e /= zero)) err es1
+    in sort roots == sort rightLocators
+
+propDecodeWithoutError :: forall c. (FiniteField c, Eq c) => ReedSolomonExample c -> Bool
+propDecodeWithoutError ReedSolomonExample {..} =
+    let encoded = encode msg pe r
+        decoded = decode encoded pe r (k+r)
+    in toPoly (V.fromList msg) == decoded
+
+propDecodeWithError :: forall c. (FiniteField c, Eq c) => ReedSolomonExample c -> Bool
+propDecodeWithError ReedSolomonExample {..} =
+    let encoded' = encode msg pe r
+        errorMsg = toPoly $ V.fromList err
+        encoded = encoded' C.+ errorMsg
+        decoded = decode encoded pe r (k+r)
+    in toPoly (V.fromList msg) == decoded
+
+specReedSolomon :: IO ()
+specReedSolomon = hspec $ do
+    describe "Reed-Solomon" $ do
+        it "generator function is correct" $ do
+            property $ propGenerator @Fr
```

Comment:
Maybe we can generalize the test suite by the base field?
In the same way how it's done in other algebraic tests (e.g. `specGroup`, `specField`)